File: js/src/answer/segment.test.ts
import { describe, expect, it } from "vitest";

import { loadCase, loadData } from "../conform/fixtures.js";
import { ABBREVIATIONS, splitClaims, TERMINATORS } from "./segment.js";

interface SegCase {
  text: string;
  claims: string[];
}

// The ADR-0009 segmentation contract: every case in
// conformance/cases/segmentation.json splits exactly as the Python reference —
// abbreviations, initials, decimals, terminator runs, CJK terminators, and the
// hard-newline rule.
/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { cases: 95 };

describe("splitClaims conformance", () => {
  const cases = loadCase<SegCase[]>("segmentation.json");

  it("vector counts are pinned", () => {
    expect({ cases: cases.length }).toEqual(EXPECTED_COUNTS);
  });

  for (const c of cases) {
    it(`splitClaims(${JSON.stringify(c.text)})`, () => {
      expect(splitClaims(c.text)).toEqual(c.claims);
    });
  }
});

describe("segmentation table", () => {
  // ADR-0010 tier 2: one canonical definition, no hand-maintained copy.
  it("matches conformance/segmentation.json exactly", () => {
    const canonical = loadData<{ terminators: string[]; abbreviations: string[] }>(
      "segmentation.json",
    );
    expect([...TERMINATORS].sort()).toEqual([...canonical.terminators].sort());
    expect([...ABBREVIATIONS].sort()).toEqual([...canonical.abbreviations].sort());
  });
});

describe("splitClaims portability traps", () => {
  // JS indexes by UTF-16 code unit; Python indexes by code point. Scanning the
  // string directly would split an astral character in half.
  it("scans by code point, not UTF-16 unit", () => {
    expect(splitClaims("𝔄 test. 𝔅 next.")).toEqual(["𝔄 test.", "𝔅 next."]);
  });

  // String.prototype.trim() strips U+FEFF, Python's str.strip() does not.
  it("trims exactly Python's whitespace set", () => {
    expect(splitClaims("﻿hello.")).toEqual(["﻿hello."]);
    expect(splitClaims("hello.")).toEqual(["hello."]);
  });
});

File: golang/answer/segment_test.go
package answer

import (
	"reflect"
	"sort"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// expectedSegmentationCases pins the size of conformance/cases/segmentation.json.
// The per-case loop below is a contract only while the case list cannot shrink.
const expectedSegmentationCases = 95

func TestSegmentationVectorCount(t *testing.T) {
	var cases []struct {
		Text   string   `json:"text"`
		Claims []string `json:"claims"`
	}
	conform.Case(t, "segmentation.json", &cases)
	if len(cases) != expectedSegmentationCases {
		t.Fatalf("segmentation.json: got %d cases, want %d", len(cases), expectedSegmentationCases)
	}
}

// TestSplitClaimsConformance is the ADR-0009 segmentation contract: every case
// in conformance/cases/segmentation.json must split exactly as the Python
// reference does — abbreviations, initials, decimals, terminator runs, CJK
// terminators, and the hard-newline rule.
func TestSplitClaimsConformance(t *testing.T) {
	var cases []struct {
		Text   string   `json:"text"`
		Claims []string `json:"claims"`
	}
	conform.Case(t, "segmentation.json", &cases)

	if len(cases) != expectedSegmentationCases {
		t.Fatalf("segmentation.json: got %d cases, want %d", len(cases), expectedSegmentationCases)
	}
	for _, c := range cases {
		got := SplitClaims(c.Text)
		want := c.Claims
		if want == nil {
			want = []string{}
		}
		if !reflect.DeepEqual(got, want) {
			t.Errorf("SplitClaims(%q)\n got  %q\n want %q", c.Text, got, want)
		}
	}
}

// TestSegmentationTableMatchesConformance pins the embedded copy to the
// canonical conformance/segmentation.json (ADR-0010 tier 2).
func TestSegmentationTableMatchesConformance(t *testing.T) {
	var canonical struct {
		Terminators   []string `json:"terminators"`
		Abbreviations []string `json:"abbreviations"`
	}
	conform.Data(t, "segmentation.json", &canonical)

	terms, abbrevs := loadSegmentationTable()

	gotTerms := make([]string, 0, len(terms))
	for r := range terms {
		gotTerms = append(gotTerms, string(r))
	}
	gotAbbrevs := make([]string, 0, len(abbrevs))
	for a := range abbrevs {
		gotAbbrevs = append(gotAbbrevs, a)
	}
	wantTerms := append([]string{}, canonical.Terminators...)
	wantAbbrevs := append([]string{}, canonical.Abbreviations...)
	for _, s := range []*[]string{&gotTerms, &gotAbbrevs, &wantTerms, &wantAbbrevs} {
		sort.Strings(*s)
	}
	if !reflect.DeepEqual(gotTerms, wantTerms) {
		t.Errorf("embedded terminators diverged:\n got  %q\n want %q", gotTerms, wantTerms)
	}
	if !reflect.DeepEqual(gotAbbrevs, wantAbbrevs) {
		t.Errorf("embedded abbreviations diverged:\n got  %q\n want %q", gotAbbrevs, wantAbbrevs)
	}
}

// TestSplitClaimsIsCodePointIndexed guards the Go-specific trap: Python indexes
// strings by code point, so the scanner must run over runes, not bytes. A
// byte-indexed scanner mangles multi-byte terminators.
func TestSplitClaimsIsCodePointIndexed(t *testing.T) {
	got := SplitClaims("従業員は開示してはならない。期間は五年である。")
	want := []string{"従業員は開示してはならない。", "期間は五年である。"}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("CJK split\n got  %q\n want %q", got, want)
	}
}
